lib/base/json.cpp

```diff
@@ -2,22 +2,255 @@
 
 #include "base/json.hpp"
 #include "base/debug.hpp"
-#include "base/namespace.hpp"
-#include "base/dictionary.hpp"
 #include "base/array.hpp"
-#include "base/objectlock.hpp"
-#include "base/convert.hpp"
+#include "base/generator.hpp"
 #include "base/utility.hpp"
-#include <bitset>
-#include <boost/exception_ptr.hpp>
-#include <cstdint>
-#include <json.hpp>
 #include <stack>
 #include <utility>
 #include <vector>
 
 using namespace icinga;
 
+constexpr uint8_t JsonEncoder::m_IndentSize;
+
+JsonEncoder::JsonEncoder(std::string& output, bool prettify)
+	: JsonEncoder{nlohmann::detail::output_adapter<char>(output), prettify}
+{
+}
+
+JsonEncoder::JsonEncoder(std::basic_ostream<char>& stream, bool prettify)
+	: JsonEncoder{nlohmann::detail::output_adapter<char>(stream), prettify}
+{
+}
+
+JsonEncoder::JsonEncoder(nlohmann::detail::output_adapter_t<char> stream, bool prettify)
+	: m_Pretty(prettify), m_Indent{0}, m_IndentStr(32, ' '), m_Writer(std::move(stream))
+{
+}
+
+/**
+ * Encodes the given value into JSON and writes it to the configured output stream.
+ *
+ * This function is specialized for the @c Value type and recursively encodes each
+ * and every concrete type it represents.
+ *
+ * @param value The value to be JSON serialized.
+ * @param yc The optional yield context for asynchronous operations. If provided, it allows the encoder
+ * to flush the output stream safely when it has not acquired any object lock on the parent containers.
+ */
+void JsonEncoder::EncodeImpl(const Value& value, boost::asio::yield_context* yc)
+{
+	switch (value.GetType()) {
+		case ValueEmpty:
+			Write("null");
+			return;
+		case ValueBoolean:
+			Write(value.ToBool() ? "true" : "false");
+			return;
+		case ValueString:
+			EncodeStringOrNumber(value.Get<String>());
+			return;
+		case ValueNumber:
+			// Make sure 0.0 is serialized as 0, so e.g. Icinga DB can parse it as int.
+			if (auto ll(static_cast<long long>(value)); ll == value) {
```

Comment:
On second thought, unless I'm wrong, that cast is actually undefined behavior if `value` is out of range for `long long`.

> If the source value can be represented exactly in the destination type, it [...] If the source value is between two representable values of the destination type, [...] Otherwise, the behavior is undefined.

-- https://en.cppreference.com/w/cpp/language/implicit_cast.html#Floating-point_conversions

Possible alternatives would be `boost::numeric_cast` (although that would involve exceptions) or maybe even just removing trailing ".0" on a string level (somewhat ugly though).